hip_data_tools/apache/cassandra.py

```diff
@@ -0,0 +1,163 @@
+def _pandas_factory(colnames, rows):
+    return pd.DataFrame(rows, columns=colnames)
+
+
+class CassandraUtil:
+    """
+    Class to connect to and then retrieve, transform and upload data from and to cassandra
+    """
+
+    def __init__(self, keyspace: str, conn: CassandraConnectionManager):
+        self.keyspace = keyspace
+        self._conn = conn
+        self._session = self._conn.get_session(self.keyspace)
+
+    @staticmethod
+    def _clean_outgoing_values(val):
+        if isinstance(val, Timestamp):
+            return val.to_pydatetime()
+        if val is NaT:
+            return None
+        return val
+
+    @staticmethod
+    def _extract_rows_from_list_of_dict(data):
+        return [tuple(dct.values()) for dct in data]
+
+    @staticmethod
+    def _extract_rows_from_dataframe(dataframe):
+        return [tuple([CassandraUtil._clean_outgoing_values(val) for val in row]) for index, row in
+                dataframe.iterrows()]
+
+    @staticmethod
+    def _prepare_batch(prepared_statement, rows) -> BatchStatement:
+        batch = BatchStatement(consistency_level=ConsistencyLevel.QUORUM)
+        for row in rows:
+            batch.add(prepared_statement, row)
+        return batch
+
+    def _cql_upsert_from_dict(self, data, table):
+        upsert_sql = f"""
+        INSERT INTO {self.keyspace}.{table} 
+        ({", ".join(data[0])}) 
+        VALUES ({", ".join(['?' for key in data[0]])});
+            """
+        print(upsert_sql)
+        return upsert_sql
+
+    def _cql_upsert_from_dataframe(self, dataframe, table):
+        upsert_sql = f"""
+        INSERT INTO {self.keyspace}.{table} 
+        ({", ".join(list(dataframe.columns.values))}) 
+        VALUES ({", ".join(['?' for key in dataframe.columns.values])});
+            """
+        print(upsert_sql)
+        return upsert_sql
+
+    def upsert_dataframe(self, dataframe: DataFrame, table: str) -> None:
+        """
+        upload all data from a DataFrame onto a cassandra table
+        Args:
+            dataframe (DataFrame): a DataFrame to upsert
+            table (str): the table to upsert data into
+            column_mapping (dict): a mapping between column names of the dataframe to cassandra
+            table. If None then the DataFrame column names that match cassandra table anme will be
+            upserted else ignored
+        Returns: None
+        """
+        prepared_statement = self._session.prepare(
+            self._cql_upsert_from_dataframe(dataframe, table))
+        batch = self._prepare_batch(prepared_statement,
+                                    self._extract_rows_from_dataframe(dataframe))
+        return self._session.execute(batch)
+
+    def create_table_from_dataframe(self, dataframe: DataFrame, table: str) -> None:
+        """
+        create a table based on a given pandas DataFrame 's schema
+        Args:
+            dataframe (DataFrame):
+            table (str):
+        Returns: None
+        """
+        pass
```

Comment:
Are we missing some code here?

Reply:
placeholder to write the code eventually when I am less tired.

Reply:
Maybe add a TODO?

Reply:
Or a github issue? Not too sure how you're tracking things. Adding incomplete methods might not be best idea without labelling them, because they'll appear in the IDE, but won't do anthing